components/automate-ui/src/app/page-components/create-user-modal/create-user-modal.component.ts

```diff
@@ -1,15 +1,140 @@
-import { Component, OnInit } from '@angular/core';
+import { Component, OnInit, Input, EventEmitter, OnChanges, SimpleChanges, OnDestroy } from '@angular/core';
+import { FormGroup, FormBuilder, Validators } from '@angular/forms';
+import { Store } from '@ngrx/store';
+import { Subject } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+
+import { NgrxStateAtom } from 'app/ngrx.reducers';
+import { Regex } from 'app/helpers/auth/regex';
+import { UsernameMapper } from 'app/helpers/auth/username-mapper';
+import { ChefValidators } from 'app/helpers/auth/validator';
+import { CreateUserPayload, CreateUser } from 'app/entities/users/user.actions';
+
+// pattern for valid usernames
+const USERNAME_PATTERN = '[0-9A-Za-z_@.+-]+';
 
 @Component({
   selector: 'app-create-user-modal',
   templateUrl: './create-user-modal.component.html',
   styleUrls: ['./create-user-modal.component.scss']
 })
-export class CreateUserModalComponent implements OnInit {
+export class CreateUserModalComponent implements OnInit, OnDestroy, OnChanges {
+  @Input() openEvent: EventEmitter<boolean>;
+
+  public createUserForm: FormGroup;
+  public conflictErrorEvent: EventEmitter<boolean>;
+  public passwordErrorEvent: EventEmitter<boolean>;
+
+  public conflictError = false;
+  public passwordError = false;
+  public modifyUsername = false; // Whether the edit Username form is open or not.
+  public creatingUser = false;
+  public visible = false;
+
+  private isDestroyed = new Subject<boolean>();
+
+  ngOnChanges(changes: SimpleChanges): void {
```

Comment:
we seem to be using `ngOnChanges` more lately. what changes get detected?

Reply:
It detects any changes on `@Input` properties. Note that whenever using `ngOnChanges` it is important to filter to the specific one you are interested in. Very much like the way we filter `@Effect` items with `ofType`, though it is a very different syntax.